libraries/AP_Math/SCurve.cpp

```diff
@@ -0,0 +1,963 @@
+/*
+ This program is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_Math/AP_Math.h>
+#include <AP_HAL/AP_HAL.h>
+#include <AP_InternalError/AP_InternalError.h>
+#include "SCurve.h"
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+#include <stdio.h>
+#endif
+
+extern const AP_HAL::HAL &hal;
+
+#define SEG_INIT        0
+#define SEG_ACCEL_MAX   4
+#define SEG_ACCEL_END   7
+#define SEG_CHANGE_END  14
+#define SEG_CONST       15
+#define SEG_DECEL_END   22
+
+// constructor
+SCurve::SCurve()
+{
+    init();
+}
+
+// initialise and clear the path
+void SCurve::init()
+{
+    jerk_time = 0.0f;
+    jerk_max = 0.0f;
+    accel_max = 0.0f;
+    vel_max = 0.0f;
+    time = 0.0f;
+    num_segs = SEG_INIT;
+    add_segment(num_segs, 0.0f, SegmentType::CONSTANT_JERK, 0.0f, 0.0f, 0.0f, 0.0f);
+    track.zero();
+    delta_unit.zero();
+    position_sq = 0.0f;
+}
+
+// set maximum velocity and re-calculate the path using these limits
+void SCurve::set_speed_max(float speed_xy, float speed_up, float speed_down)
+{
+    // segment accelerations can not be changed after segment creation.
+    const float track_speed_max = kinematic_limit(delta_unit, speed_xy, speed_up, fabsf(speed_down));
+
+    if (is_equal(vel_max, track_speed_max)) {
+        // new speed is equal to current speed maximum
+        return;
+    }
+
+    if (is_zero(vel_max) || is_zero(track_speed_max)) {
+        // new or original speeds are set to zero
+        return;
+    }
+    vel_max = track_speed_max;
+
+    // Check path has been defined
+    if (num_segs != segments_max) {
+        return;
+    }
+
+    if (time >= segment[SEG_CONST].end_time) {
+        return;
+    }
+
+    // re-calculate the s-curve path based on update speeds
+
+    const float Pend = segment[SEG_DECEL_END].end_pos;
+    float Vend = MIN(vel_max, segment[SEG_DECEL_END].end_vel);
+    uint8_t seg;
```

Comment:
it's just an index used as an input/output to the add_segments_xxx methods.  I've reduced the scope so that "seg" is declared closer to where it's used.